tools/fraken/main.go

```diff
@@ -0,0 +1,451 @@
+		return s.compile()
+	}
+
+	return nil
+}
+
+func filesystemScan(wait chan struct{}, c chan *Detection) {
+	if _, err := os.Stat(*scanPathFlag); err != nil {
+		log.Printf("cannot scan %v: %v\n", *scanPathFlag, err)
+		close(c)
+		return
+	}
+	var wg sync.WaitGroup
+	err := filepath.Walk(*scanPathFlag, func(filePath string, fileInfo os.FileInfo, err error) error {
+		if !fileInfo.Mode().IsRegular() {
+			return nil
+		}
+
+		wg.Add(1)
+		wait <- struct{}{}
+		go func() {
+			matches, _ := scanFile(scanner, filePath, fileInfo)
+			for _, match := range matches {
+				var description, reference string
+				score := 50
+				for _, m := range match.Metas {
+					if m.Identifier == "score" {
+						score = m.Value.(int)
+					}
+					if strings.HasPrefix(m.Identifier, "desc") {
+						description = m.Value.(string)
+					}
+					if m.Identifier == "reference" || strings.HasPrefix(m.Identifier, "report") {
+						reference = m.Value.(string)
+					}
+					if m.Identifier == "context" {
+						v := strings.ToLower(m.Value.(string))
+						if v == "yes" || v == "true" || v == "1" {
+							score = 0
+						}
+					}
+				}
+				c <- newDetection(filePath, match.Rule, description, reference, score)
+			}
+			<-wait
+			wg.Done()
+		}()
+		return nil
+	})
+	if err != nil {
+		log.Printf("error walking dir: %v\n", err)
+	}
+	wg.Wait()
+	close(c)
+}
+
+func scanFile(s Scanner, filePath string, fileInfo os.FileInfo) (yara.MatchRules, error) {
+	var matches yara.MatchRules
+
+	ys, err := yara.NewScanner(s.Rules)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load scanner: %v", err)
+	}
+
+	// Fill the variables
+	ys.DefineVariable("filepath", filePath)
+	ys.DefineVariable("filename", fileInfo.Name())
+	ys.DefineVariable("extension", filepath.Ext(filePath))
+
+	stat := fileInfo.Sys().(*syscall.Stat_t)
+	owner, err := user.LookupId(strconv.FormatUint(uint64(stat.Uid), 10))
+	if err == nil {
+		ys.DefineVariable("owner", owner.Username)
+	}
+
+	ys.DefineVariable("filetype", getFileTypes(filePath))
+
+	// Scan the file.
+	err = ys.SetCallback(&matches).ScanFile(filePath)
+
+	if err != nil {
+		return matches, err
+	}
+
+	// Return any results.
+	return matches, nil
+}
+
+func init() {
+	scanPathFlag = flag.String("folder", "", "Specify a particular folder to be scanned")
+	rulePathFlag = flag.String("rules", "", "Specify a particular path to a file or folder containing the Yara rules to use")
+	magicPathFlag = flag.String("magic", "misc/file-type-signatures.txt", "A path under the rules path that contains File Magics")
+	yaraRulesFlag = flag.String("extrayara", "", "Any additional Yara rules to be used")
+	flag.Parse()
+}
+
+func main() {
+	if *scanPathFlag == "" || *rulePathFlag == "" {
+		log.Println("Usage: fraken -folder <path to scan> -rules <path to rules> [-magic <path to magics>] [-extrayara <path to file>]")
+		os.Exit(1)
+	}
+	err := scanner.init()
+	if err != nil {
+		log.Printf("error initialising Yara engine: %v\n", err)
```

Comment:
You can just use log.Fatalf() which calls os.Exit

Reply:
Done